src/x86/misc.rs

```diff
@@ -0,0 +1,150 @@
+//! Miscelaneous x86 intrinsics available on all hosts.
+//!
+//! TODO: figure out a more discoverable name.
+
+#[cfg(test)]
+use stdsimd_test::assert_instr;
+
+/// Result of the `cpuid` instruction.
+pub struct CpuidResult {
+    /// EAX register.
+    pub eax: u32,
+    /// EBX register.
+    pub ebx: u32,
+    /// ECX register.
+    pub ecx: u32,
+    /// EDX register.
+    pub edx: u32,
+}
+
+/// `cpuid` instruction.
+///
+/// The [CPUID Wikipedia page][wiki_cpuid] contains how to query which
+/// information using the `eax` and `ecx` registers, and the format in
+/// which this information is returned in `eax...edx`.
+///
+/// The `has_cpuid()` intrinsics can be used to query whether the `cpuid`
+/// instruction is available.
+///
+/// The definitive references are:
+/// - [Intel 64 and IA-32 Architectures Software Developer's Manual Volume 2:
+///   Instruction Set Reference, A-Z][intel64_ref].
+/// - [AMD64 Architecture Programmer's Manual, Volume 3: General-Purpose and
+///   System Instructions][amd64_ref].
+///
+/// [wiki_cpuid]: https://en.wikipedia.org/wiki/CPUID
+/// [intel64_ref]: http://www.intel.de/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-instruction-set-reference-manual-325383.pdf
+/// [amd64_ref]: http://support.amd.com/TechDocs/24594.pdf
+#[inline(always)]
+#[cfg_attr(test, assert_instr(cpuid))]
+pub unsafe fn cpuid(eax: u32, ecx: u32) -> CpuidResult {
+    let mut r = ::std::mem::uninitialized::<CpuidResult>();
+    asm!("cpuid"
+         : "={eax}"(r.eax), "={ebx}"(r.ebx), "={ecx}"(r.ecx), "={edx}"(r.edx)
+         : "{eax}"(eax), "{ecx}"(ecx)
+         : :);
+    r
+}
+
+/// Reads EFLAGS.
+#[cfg(target_arch = "x86")]
+#[inline(always)]
+pub unsafe fn pushfd() -> u32 {
+    let eflags: u32;
+    asm!("pushfd; popl $0" : "=r"(eflags) : : : "volatile");
+    eflags
+}
+
+/// Write EFLAGS.
+#[cfg(target_arch = "x86")]
+#[inline(always)]
+pub unsafe fn popfd(eflags: u32) {
+    asm!("pushl $0; popfd" : : "r"(eflags) : "cc", "flags" : "volatile");
+}
+
+/// Does the host support the `cpuid` instruction?
+#[inline(always)]
+pub fn has_cpuid() -> bool {
+    #[cfg(target_arch = "x86_64")] {
+        true
+    }
+    #[cfg(target_arch = "x86")] {
+        // On `x86` the `cpuid` instruction is not always available.
+        // This follows the approach indicated in:
+        // http://wiki.osdev.org/CPUID#Checking_CPUID_availability
+        unsafe {
+            // Read EFLAGS:
+            let eflags: u32 = pushfd();
+
+            // Invert the ID bit in EFLAGS:
+            let eflags_mod: u32 = eflags | 0x0020_0000;
+
+            // Store the modified EFLAGS (ID bit may or may not be inverted)
+            popfd(eflags_mod);
+
+            // Read EFLAGS again:
+            let eflags_after: u32 = pushfd();
+
+            // Check if the ID bit changed:
+            eflags_after != eflags
```

Comment:
@gnzlbg 
Have you checked what exact instruction sequence is produced?
OSDev wiki warns: "Note: Implementing this routine in for example C can lead to issues, because the compiler may change EFLAGS at any time."
(Any intermediate arithmetic operations can change `EFLAGS`.)

Reply:
Yes I have, and the build should be green now :)

Reply:
It produces this in release:

```asm
ex:
	pushfl
	popl	%eax
	movl	%eax, %ecx
	orl	$2097152, %ecx
	pushl	%ecx
	popfl
	pushfl
	popl	%ecx
	cmpl	%eax, %ecx
	setne	%al
	retl
```

and this in debug:

```asm
	ex:
	subl	$12, %esp
	pushfl
	popl	%eax
	movl	%eax, 4(%esp)
	movl	4(%esp), %eax
	movl	%eax, %ecx
	orl	$2097152, %ecx
	pushl	%ecx
	popfl
	pushfl
	popl	%ecx
	movl	%ecx, 8(%esp)
	cmpl	%eax, 8(%esp)
	setne	%dl
	movb	%dl, 3(%esp)
	movb	3(%esp), %al
	andb	$1, %al
	movzbl	%al, %eax
	addl	$12, %esp
	retl
```

not "perfect" but good enough, plus I got to factor `pushfd` and `popf` out :)